src/main/java/com/epam/aidial/core/service/ApplicationService.java

```diff
@@ -279,4 +489,420 @@ private void prepareApplication(ResourceDescription resource, EtagHeader etag,
+                    if (existing == null || !Objects.equals(existing.getFunction(), function)) {
+                        throw new IllegalStateException("Application function has been updated");
+                    }
+
+                    Application.Function.Status status = (function.getStatus() == Application.Function.Status.STOPPING)
+                            ? Application.Function.Status.STOPPED
+                            : Application.Function.Status.FAILED;
+
+                    function.setStatus(status);
+                    function.setError(status == Application.Function.Status.FAILED ? error : null);
+
+                    existing.setFunction(function);
+                    return ProxyUtil.convertToString(existing);
+                });
+            }
+
+            pendingApplications.remove(resource.getUrl());
+            return null;
+        } catch (Throwable e) {
+            log.warn("Failed to terminate application: {}", resource.getUrl(), e);
+            throw e;
+        }
+    }
+
+    private void deleteApplicationImage(Application.Function function) {
+        callController(HttpMethod.DELETE, "/v1/image/" + function.getId(),
+                request -> null,
+                (response, body) -> convertServerSentEvent(body, EmptyResponse.class));
+    }
+
+    private void deleteApplicationDeployment(Application.Function function) {
+        callController(HttpMethod.DELETE, "/v1/deployment/" + function.getId(),
+                request -> null,
+                (response, body) -> convertServerSentEvent(body, EmptyResponse.class));
+    }
+
+    // endregion
+
+    private void copyFolder(String sourceFolderUrl, String targetFolderUrl) {
+        ResourceDescription sourceFolder = ResourceDescription.fromAnyUrl(sourceFolderUrl, encryptionService);
+
+        String token = null;
+        do {
+            ResourceFolderMetadata folder = resourceService.getFolderMetadata(sourceFolder, token, PAGE_SIZE, true);
+            if (folder == null) {
+                throw new IllegalStateException("Source folder is empty");
+            }
+
+            for (MetadataBase item : folder.getItems()) {
+                String sourceFileUrl = item.getUrl();
+                String targetFileUrl = targetFolderUrl + sourceFileUrl.substring(sourceFolder.getUrl().length());
+
+                ResourceDescription sourceFile = ResourceDescription.fromAnyUrl(sourceFileUrl, encryptionService);
+                ResourceDescription targetFile = ResourceDescription.fromAnyUrl(targetFileUrl, encryptionService);
+
+                if (!resourceService.copyResource(sourceFile, targetFile)) {
+                    throw new IllegalStateException("Can't copy function source file: " + sourceFileUrl);
+                }
+            }
+
+            token = folder.getNextToken();
+        } while (token != null);
+    }
+
+    private void deleteFolder(String folderUrl) {
```

Comment:
Let's move the method to ResourceService

Reply:
done